DOCS: improved documentation

antoinecollet5 opened the pull request:
Several changes I propose for the documentation:

Documentation:

-    Use the modern pydata template rather than rtd (used by mainstream packages such as numpy and scipy)
-    Include the readme in the doc for a nice rendering and avoid duplication
-    Make links between the bibliography and the docstrings using the :cite: directive.
-    Improved docstrings for SIES and ES (differentiate class attributes for method parameters, detail inversion types) .
-    Better API section
-    Add makefiles for documentation build (both unix and windows) because tox does not work offline.
-    Add badges in readme for better visibility

tommyod commented:
Will look over this more closely soon. :) We just merged a big overhaul of SIES, leading to some conflicts unfortunately. You can probably just merge main and checkout ``--theirs` to solve the conflicts without too much trouble.

tommyod commented:
Hi. Fantastic work as always! Mostly looks good to me. A few comments:

- You'll have to merge in our recent changes unfortunately.
- We want to avoid makefiles. Can you elaborate on why they are needed if they are? If they are not needed, then we should remove them from the PR.

I'll also try to run it locally before we merge, just to see what the docs look like. But I'll wait until you have merged main :) Good stuff. Thanks for the effort. Greatly appreciate it.

antoinecollet5 commented:
@tommyod thanks for the feedbacks :) I found a bit of time to solve the conflicts and adapt to the new interface. Note that I removed the example in the readme since it is now outdated.

Regarding the  Makefiles, I just removed them.

Wait for your feedback about the html rendering. Note that I did not take care of the pdf export.

tommyod commented on docs/source/_templates/autosummary/module.rst:
```diff
@@ -0,0 +1,61 @@
+{{ fullname | escape | underline}}
```
Where is this file from? It is auto-generated from some system? Could you add a reference/source as a comment on the top?

antoinecollet5 replied:
The three files in the _template/autosummary/ folder actually define how classes, modules etc. render with autosummary. It is not automatically generated. I can add a readme at the root of the folder if you judge is necessary.

tommyod replied:
If you copied it from some other repo or source, then please add a line saying so. If you wrote it yourself, then no reason to add anything.

antoinecollet5 replied:
These three files have been modified from the default templates provided by autosummary. See https://sphinx-rtd-trial.readthedocs.io/en/latest/ext/autosummary.html, section "Customizing templates".
The original files can be found here: https://github.com/sphinx-doc/sphinx/tree/master/sphinx/ext/autosummary/templates/autosummary

I will add the reference in each file.

tommyod commented on docs/source/conf.py:
```diff
@@ -1,26 +1,258 @@
+#!/usr/bin/env python
+#
+# pyesmda documentation build configuration file, created by
```
Remove reference to `pyesmda` please :) Perhaps we should run this setup again, it if it was run in 2017 (many years ago)?

antoinecollet5 replied:
Fixed. I cleaned up the conf.py

tommyod commented on docs/source/conf.py:
```diff
@@ -1,26 +1,258 @@
 from subprocess import check_output
 
+import sphinx.util.osutil
+from sphinx.ext.napoleon.docstring import GoogleDocstring
```
We have decided on NumPy docstrings, so I think all references to Google docstrings can go away.

antoinecollet5 replied:
These stuff were simply to rename the "Variable" section to "Attributes" but since you think exposing attributes might not be a good idea, this is not required.

tommyod commented on docs/source/conf.py:
```diff
@@ -1,26 +1,258 @@
-
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
 extensions = [
```
Anything here that we do not need? This list grew quite a bit.

antoinecollet5 replied:
There were duplicates. I cleaned it.

tommyod commented on docs/source/conf.py:
```diff
@@ -1,26 +1,258 @@
+    "jupyter_sphinx",
     "numpydoc",
+    "m2r2",  # compatibility with markdown
+    "myst_nb",  # to include jupyter nteboo
```
I note that https://github.com/spatialaudio/nbsphinx has been replaced with https://github.com/executablebooks/MyST-NB. I don't mind, if we have a reason for it.

antoinecollet5 replied:
Well... I did not manage to make nbsphinx work. Don't really understand why. I usually rely on this lib. Myst works fine too. I added some config so that the maths render correctly.

tommyod commented on docs/source/conf.py:
```diff
@@ -1,26 +1,258 @@
-bibtex_bibfiles = ["refs.bib"]
+
+suppress_warnings = [
+    "nbsphinx",
```
No need to suppress warnings if it's not used any more?

antoinecollet5 replied:
Good catch :)

tommyod commented on docs/source/conf.py:
```diff
@@ -1,26 +1,258 @@
+    return self._format_fields("Keys", self._consume_fields())
+
+
+GoogleDocstring._parse_keys_section = parse_keys_section
```
Remove google stuff, since we use NumPy docstrings.

antoinecollet5 replied:
Removed.

tommyod commented on docs/source/index.rst:
```diff
@@ -1,6 +1,4 @@
-===========================
-Iterative Ensemble Smoother
-===========================
+.. mdinclude:: ../../README.md
```
Very cool! I love the re-use of the README.

antoinecollet5 replied:
Cool :)

tommyod commented on src/iterative_ensemble_smoother/sies.py:
```diff
@@ -17,50 +17,61 @@
+    The class attributes follow the paper nomenclature.
 
-    Parameters
+    Attributes
```
I think we should avoid "exposing" attributes in the docs. They are meant for internal use by the class and could in principle change names. Users should interact via the API, not via attributes (unlike e.g. sklearn models, where users are also encouraged to get e.g. `coefs_` from a linear model).

At least that is my thinking. I think exposing attributes leads to more noise, more stuff to maintain, more backwards compatibility issues, etc.

antoinecollet5 replied:
Ok I removed it. And you are right, this is anyway  to discuss in another PR anyway.

tommyod reviewed (commented):
Overall very excited about this. New theme looks great. Many improvements here. I do have some comments however :)

New theme looks good:

![image](https://github.com/equinor/iterative_ensemble_smoother/assets/10076072/eb58ecfc-c551-41c5-ac17-8f0fb0e6859b)


But LaTeX does not render:

![image](https://github.com/equinor/iterative_ensemble_smoother/assets/10076072/3d36f0c0-fed6-4a3e-9d26-dba355b34629)

Adding `"sphinx.ext.mathjax"` to the `extensions` in `conf.py` did not help me locally.

- Could you take a look at LaTeX? Does it work for you?
- I added some comments along the way, please take a look.
- In the future, even smaller PRs is good. For instance, I am not so sure about documenting attributes, but willing to discuss it. However, it's not really related to the issue of updating sphinx etc - so it would've been even better to have a PR for documentation changes (unless they are strictly required to make sphinx run).

antoinecollet5 commented:
@tommyod I had a quick look at latex and indeed it does not render. I propose you to see that in another issue/PR ? And to discuss whether the latexpdf export is really needed ?

tommyod commented:
> And to discuss whether the latexpdf export is really needed ?

The latexpdf is not needed. But math should render in the HTML. For me it does not render correctly in HTML (see screenshot above, same now as before). Does it render in the HTML docs for you?

I see that [m2r2 expects inline math as](https://github.com/CrossNox/m2r2/issues/20) `` `$e=mc^2$` ``, with the backticks. We use `$e=mc^2$` in our markdown. For blocks of math we use `$$e=mc^2$$` in the notebooks. I am not sure what m2r2 excepts here, or if it supports it at all.

> I propose you to see that in another issue/PR ?

I know it's tedious, but I'd rather get HTML math to work in this PR. I don't want to merge PRs that improve some things, but regress on others, and think that we'll solve it in a future PR. This kind of thinking has a tendency to lead to messy code, especially when it's not obvious what the fix is, who will dedicate time to fix it, etc.

antoinecollet5 commented:
@tommyod normally, math should now render correctly in the html export. Have you tried since yesterday's changes ?

tommyod commented:
Maybe I did something wrong. Let me try again and report back ... :)

tommyod commented:
Fresh try:

```
git remote add antoine https://github.com/antoinecollet5/iterative_ensemble_smoother.git
git fetch --all
git switch feature/docs
pip install -e '.[dev,doc]'
python -m ipykernel install --user
sphinx-build -c docs/source/ -b html docs/source/ docs/build/html -W
firefox docs/build/html/index.html
```

And it works and looks great. :+1:

tommyod reviewed (approved):
LGTM. Thanks again for the work @antoinecollet5 . Very much appreciated as always!
